pkg/storage/unified/resource/resource.proto

```diff
@@ -0,0 +1,480 @@
+  string origin = 4;
+
+  // Path on the origin
+  string path = 5;
+
+  // Verification hash from the origin
+  string hash = 6;
+
+  // Change time from the origin
+  int64 timestamp = 7;
+}
+
+message OriginResponse {
+  repeated ResourceOriginInfo items = 1;
+
+  // More results exist... pass this in the next request
+  string next_page_token = 2;
+
+  // ResourceVersion of the list response
+  int64 resource_version = 3;
+}
+
+message HealthCheckRequest {
+  string service = 1;
+}
+
+message HealthCheckResponse {
+  enum ServingStatus {
+    UNKNOWN = 0;
+    SERVING = 1;
+    NOT_SERVING = 2;
+    SERVICE_UNKNOWN = 3;  // Used only by the Watch method.
+  }
+  ServingStatus status = 1;
+}
+
+
+//----------------------------
+// Blob Support
+//----------------------------
+
+message PutBlobRequest {
+  enum Method {
+    // Use the inline raw []byte
+    GRPC = 0;
+
+    // Get a signed URL and PUT the value 
+    HTTP = 1;
+  }
+
+  // The resource that will use this blob
+  // NOTE: the name may not yet exist, but group+resource are required
+  ResourceKey resource = 1;
+
+  // How to upload
+  Method method = 2;
+
+  // Content type header
+  string content_type = 3;
+
+  // Raw value to write
+  // Not valid when method == HTTP
+  bytes value = 4;
+}
+
+message PutBlobResponse {
+  // Status code
+  StatusResult status = 1;
+
+  // The blob uid.  This must be saved into the resource to support access
+  string uid = 2;
+
+  // The URL where this value can be PUT
+  string url = 3;
+
+  // Size of the uploaded blob
+  int64 size = 4;
+
+  // Content hash used for an etag
+  string hash = 5;
+
+  // Validated mimetype (from content_type)
+  string mime_type = 6;
+
+  // Validated charset (from content_type)
+  string charset = 7;
+}
+
+message GetBlobRequest {
+  ResourceKey resource = 1;
+
+  // The new resource version
+  int64 resource_version = 2;
+
+  // Do not return a pre-signed URL (when possible)
+  bool must_proxy_bytes = 3;
+}
+
+message GetBlobResponse {
+  // Status code sent on errors
+  StatusResult status = 1;
+
+  // (optional) When possible, the system will return a presigned URL
+  // that can be used to actually read the full blob+metadata
+  // When this is set, neither info nor value will be set
+  string url = 2;
+
+  // Content type 
+  string content_type = 3;
+
+  // The raw object value
+  bytes value = 4;
+}
+
+
+// This provides the CRUD+List+Watch support needed for a k8s apiserver
+// The semantics and behaviors of this service are constrained by kubernetes
+// This does not understand the resource schemas, only deals with json bytes 
+// Clients should not use this interface directly; it is for use in API Servers
+service ResourceStore {
+  rpc Read(ReadRequest) returns (ReadResponse);
+  rpc Create(CreateRequest) returns (CreateResponse);
+  rpc Update(UpdateRequest) returns (UpdateResponse);
+  rpc Delete(DeleteRequest) returns (DeleteResponse);
+  rpc List(ListRequest) returns (ListResponse);
```

Comment:
ListResponse will be pretty big. We should try to make this a stream of Items instead of returning all the items in one single response.

Reply:
what does the etcd flavor do?

Given that it translates to paged HTTP requests in the end -- i don't think the stream is really necessary.  However we will need some logic like we implemented for dashboards that makes the pages bound by byte size, not just count.

🤔 the answer here may really depend on how much filtering is possible how far downs the stack.  If we stream, does each item come with a continue token?

Reply:
Ok...ETCD itself doesn't stream (see [Range grpc request](https://github.com/etcd-io/etcd/blob/main/api/etcdserverpb/rpc.proto#L39-L43)). I guess with pagination it's OK as long at the resources are not too big.